### PR TITLE
perf(deepseek-v4): widen sparse compress cache launch for large ratios

### PR DESCRIPTION
**Pull request by Xiangyi1996**
## Summary

Launch change for the fused sparse compress cache insert: 16 warps when `compress_ratio >= 128` on validated targets (NVIDIA sm100 only, gated by `_wide_compress_launch_supported()`; all other targets including ROCm keep 4 warps), plus launch-config and GPU parity tests.

Boundary tokens reduce a (1+overlap)*ratio x HEAD window inside one CTA: 128 rows on the production HCA path (overlap=False); overlapping configurations reduce 256. At 4 warps that reduction is severely under-parallelized.

## Evidence

- Microbench (B200, graph replay): boundary path 450-730us -> 143-190us (3-3.8x); prefill shapes ratio=128 M=8192: 905 -> 281us; ratio=4 path unchanged (kept at 4 warps, where 16 is slightly slower).
- p8 exact decode trace (same protocol pair): kernel-sum 2.184 -> 0.561 ms/step (-74%, 36.2 -> 9.3 us/call, 60 calls/step).
- Cadence in that trace improved 0.56 ms/step, but the traced tree also contained the SWA sanitize hoist (#614); the cadence change is not isolated to this patch.
- Endpoint gain is not certified: expected magnitude (~2%) is below current arm-level measurement noise.
- Numerics: the wide launch can reorder fp32 reductions before FP8/BF16 quantization. GPU parity tests (production HCA shape ratio=128/overlap=False, M in {16,32,64,2048}, multi-seed) compare every byte of every WRITTEN row at 4 vs 16 warps — FP8 values, encoded scales, and the RoPE BF16 tail as separate regions — including a partial-window case where block_table_base_offsets masks the leading window rows. On B200 all written rows are byte-identical (asserted with torch.equal).
- Microbench overlap note: the 3-3.8x boundary-path numbers were measured on a synthetic overlap=True (256-row) shape; the production claim is carried by the exact trace (36.2 -> 9.3 us/call).
- Scope: 16-warp launch is gated to NVIDIA sm100; ROCm and other targets keep 4 warps by construction.

Tests: 5 launch-config tests (wide/narrow launch per ratio, unsupported-target fallback, one-program-per-token grid, empty batch) + 6 GPU parity cases (sm100-gated). The 16-warp gate is cached per device (mixed-architecture processes pick per-GPU).

## Update: interaction with the newer #563 prefill-fusion commits (2026-07-10)

#563's `HCA_BLOCK_TABLE_CAPACITY` constexpr touches the same kernel. Re-microbenched on a stack including those commits (B200, production HCA shape, boundary tokens):

| M | 4 warps | 16 warps | speedup |
| ---: | ---: | ---: | ---: |
| 16-64 (constexpr path) | 142-191 us | 63-73 us | 2.2-2.6x |
| 16-64 (legacy path) | 145-192 us | 43-47 us | 3.3-4.0x |
| 2048 prefill | 221-222 us | 52-92 us | 2.4-4.3x |

- The 16-warp win survives the newer #563 base (2.2-4x on the boundary reduction); absolute per-step savings there are smaller than on the base this PR was first traced on, so the endpoint claim remains "kernel-level only".
- Noted for joint tuning: at 16 warps the pow2-bucket constexpr path measures slower than the legacy runtime-bound path (63 vs 44 us); capacity specialization and warp count interact.
- This PR stays main-based; if #563 merges first, the known conflict in this file (`max_gather_len`, `HCA_BLOCK_TABLE_CAPACITY`) resolves by applying this one-line launch change on top (verified resolution exists).

### CODE REVIEW

File: tokenspeed-kernel/test/test_deepseek_v4_sparse_compress_parity_gpu.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+BF16 tail, each region separately. Token classes cover early-return (non-boundary), the first-window boundary
+(start == 0), full history, and — via block_table_base_offsets — partially
+masked windows crossing the base logical page.
+"""
+
+from __future__ import annotations
+
+import pytest
+import torch
+import triton
```
[the remark this answers is not in the document]
Comment (Xiangyi1996) [reply tokenspeed-kernel/test/test_deepseek_v4_sparse_compress_parity_gpu.py]: Fixed: triton is now imported through the tokenspeed_kernel._triton shim.

File: tokenspeed-kernel/test/test_deepseek_v4_sparse_compress_parity_gpu.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+@pytest.mark.parametrize("m", [16, 64])
+def test_sparse_compress_parity_partial_window_with_base_offsets(m):
+    # Production passes state_base_logical_page; when the reduction window
+    # crosses the base offset, leading rows resolve to table_idx < 0 and are
+    # masked. The masked-softmax reduction is exactly where a warp-count
+    # change reorganizes the computation, so compare 4 vs 16 warps here too.
+    # pos[1::4] = 1023 boundaries: rows 896..959 fall below base page 15 and
+    # are masked, rows 960..1023 stay valid -> a genuine partial window.
+    base = torch.full((4,), 15, device="cuda:0", dtype=torch.int32)
+    slots = _written_slots(m)
```
[the remark this answers is not in the document]
Comment (Xiangyi1996) [reply tokenspeed-kernel/test/test_deepseek_v4_sparse_compress_parity_gpu.py]: Fixed: the partial-window test now compares only the pos=1023 boundaries (genuine partial windows); the all-masked pos=127 rows are excluded from byte comparison with a comment explaining why.

**Comment by Xiangyi1996** [comment]
CI triage for the previous run:

- **ut-tokenspeed-kernel (gb200-1gpu)**: real finding, fixed. At M=2048 the 4-vs-16-warp reduction reorder flipped a signed zero in one FP8 output byte (0x00 vs 0x80 — byte-different, value-identical; dequant max abs was exactly 0.0). The parity test now compares FP8 and RoPE BF16 regions by decoded value (signed zeros compare equal) while scale bytes remain byte-compared. B200 re-verified 6/6.
- **eval-qwen3.5 / perf-kimi (b300-4gpu)**: unrelated to this PR — neither model family invokes the DeepSeek-V4 sparse compress cache kernel, and this one-line launch-width change is V4-HCA-specific and sm100-gated. These jobs show intermittent failures on other recent runs as well.

**Comment by Xiangyi1996** [comment]
GB200 round 2: with value-level comparison the remaining diff is a genuine FP8 rounding-tie flip from reduction reorder — 1 value in ~4.6e5 written values differing by one quantization step (1.5e-5). The parity test now bounds mismatch fraction (<1e-4) and dequantized abs/rel error (<0.25) instead of requiring bit equality across warp schedulers, matching the original test design intent. B200 6/6 re-verified.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/triton/deepseek_v4.py [hunk 3]
```diff
@@ -402,6 +404,28 @@ def _deepseek_v4_fused_sparse_compress_cache_kernel(
...
+def _wide_compress_launch_supported(device: "torch.device | int | None") -> bool:
+    """Whether the 16-warp sparse-compress launch is validated for a GPU.
+
+    Args:
+        device: Device the kernel launches on (the state cache's device);
+            results are cached per device so mixed-architecture processes
+            pick the right launch width per GPU.
+
+    Returns:
+        True only on NVIDIA sm100 (B200), where the wide launch was
```
[the remark this answers is not in the document]
Comment (Xiangyi1996) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/triton/deepseek_v4.py:417]: I don't have B300 numbers yet and the occupancy tradeoff may differ, so I'd rather extend the gate to (10,3) in a follow-up after running the same sweep on a B300 runner. The per-device gate is structured so that's a one-line change. The wide launch is only enabled where it was actually measured — sm100 (B200/GB200), where the ratio=128 boundary microbench showed 2.2–4x. On sm103 the gate falls back to the existing 4-warp launch, so behavior there is bit-identical to main.
